### PR TITLE
New event Space Law Change

### PR DESCRIPTION
**Pull request by chavonadelal**
## About the PR

Added a new event Space Law Change. It is announced to the station that something has become illegal. To confirm that the law has been changed, all faxes receive a fax with the same text and the stamp of Central Command. After time for adaptation has passed, a confirmation announcement occurs that the change in law has indeed come into force.
## Why / Balance

The event creates new role-playing situations and also adds flavor to the round.. The drop chance is very low, similar to the Ion Storm event. And the meaning is similar to the Ion Storm event only for people.
Everything can be edited - the frequency of event, what exactly will be illegal. I would be glad to see such an event appear in SS14. There is exactly the same event in SS13, which is what I was inspired by when I did it.
This is my first serious implementation of programming knowledge. I learned C# specifically to implement this event in SS14, so I would be very happy if my pull request is accepted. 👉👈🥺
## Technical details

The event fully supports localization. First, a random message is generated using a list of alert part localization keys for a given event. This message is then announced and sent to all fax machines stamped by Central Command. After the time for adaptation has expired, a message appears confirming that the law has indeed been changed. To prevent the message from being repeated, a check is made for laws that have already been changed. Suddenly the event has already happened before. At first I wanted the fax to only go to the captain's and central command's faxes, but I couldn't do that, so it's being sent out to everyone. I think it's good. More players will have access to documented evidence that the law has indeed been changed.
## Media

image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl: to4no_fix
- add: Added a new event Space Law Change. It is announced to the station that something has become illegal. To confirm that the law has been changed, all faxes receive a fax with the same text and the stamp of Central Command. After time for adaptation has passed, a confirmation announcement occurs that the change in law has indeed come into force.

### CODE REVIEW

**Comment by chavonadelal** [comment]
I did not find any errors or fatal errors during testing. Everything worked and should work.

**Comment by Clone0401** [comment]
These papers should probably be locked and un-editable with the cybersun pen as they use the Centcomm stamp

**Comment by Cojoke-dot** [comment]
> These papers should probably be locked and un-editable with the cybersun pen as they use the Centcomm stamp.
> 
> But besides that I really like this and think it would be pretty damn funny to have every so often 

Nah, if they are editable it would be much more fun. Tricking the captain with a document sent from centcom about law changes could be some fun rp.

**Comment by chavonadelal** [comment]
> > These papers should probably be locked and un-editable with the cybersun pen as they use the Centcomm stamp.
> > But besides that I really like this and think it would be pretty damn funny to have every so often
> 
> Nah, if they are editable it would be much more fun. Tricking the captain with a document sent from centcom about law changes could be some fun rp.

This is the essence, like the essence of the entire event - to add fun to the round :) You can experiment with options from banning alcohol to banning wearing clothes, for example, with the color green :)))) But I settled on something more or less concrete, since the event is new and it’s worth starting slowly.

**Comment by JIPDawg** [comment]
I don't really have an opinion one way or another, but a suggestion. Maybe use one of these form's for the fax paper to make it feel more official? https://github.com/Moomoobeef/ss14-forms-txt/tree/master/General-Headers-Footers

**Comment by Hrosts** [comment]
A lot of the text in the announcements is repeated. If it's possible to separate the "Due to recent changes..." template from the law subjects (hats, knifes, not greeting), you should probably do that.

**Comment by chavonadelal** [comment]
> A lot of the text in the announcements is repeated. If it's possible to separate the "Due to recent changes..." template from the law subjects (hats, knifes, not greeting), you should probably do that.

This is problematic to do while maintaining full accessibility to localization.  At least I don’t know how to implement this without losing some of the localization capabilities.  Moreover, the event is sensitive and I think the opportunity to completely rewrite some version of the message will be useful if someone wants to seriously change something.  It will be enough to change this simply in localization.

**Comment by Djungelskog2** [comment]
I like the idea, and the idea of captains creating custom laws but theres 2 issues i find
1: Theres nowhere to check these new laws unless relayed through constant announcements or people telling you
2: people already dont follow basic contraband rules and sec does nothing to enforce the law unless a literal murder has been committed.

**Comment by Brandon-Huu** [comment]
> sec does nothing to enforce the law unless a literal murder has been committed.

That is an admin issue. They need to ban these security players for refusing to do their jobs.
There is already a rule for this.

**Comment by Djungelskog2** [comment]
I guess your right but it feels really silly to have to ahelp some third party because someone decided to wear some pajamas

**Comment by chavonadelal** [comment]
> I like the idea, and the idea of captains creating custom laws but theres 2 issues i find 1: Theres nowhere to check these new laws unless relayed through constant announcements or people telling you 2: people already dont follow basic contraband rules and sec does nothing to enforce the law unless a literal murder has been committed.



> I like the idea, and the idea of captains creating custom laws but theres 2 issues i find 1: Theres nowhere to check these new laws unless relayed through constant announcements or people telling you 2: people already dont follow basic contraband rules and sec does nothing to enforce the law unless a literal murder has been committed.

As for the impossibility of verification, absolutely all faxes at the station will receive a confirmation document, especially since, according to the meaning of the message, before the law comes into force, players have 15 minutes to adapt to the change in the law and receive information about it.
Of course, there is also the option of creating news about this, but it is easily erased and even easier to falsify. Although it would be very convenient for obtaining information.

**Comment by Simyon264** [comment]
> Nah, if they are editable it would be much more fun. Tricking the captain with a document sent from centcom about law changes could be some fun rp.

Many people see papers stamped by Centcom as messages by ADMINS. If the stamped paper said "kill X", the chance is high that someone does it even if it goes 100% against rules.

**Comment by Djungelskog2** [comment]
laminate the paper to make it un-editable

**Comment by chavonadelal** [comment]
> laminate the paper to make it un-editable

@Djungelskog2  I'm not sure it's worth doing, but I don't know how to do it either. Could you tell? If the reviewer decides that it needs to be made uneditable

**Comment by Djungelskog2** [comment]
I just said that as a flavourful way that un-editable papers could be implemented, im guessing a new texture would be overlapped onto the paper but yea thats all I got

**Comment by Clone0401** [comment]
> What papers in the game are un-editable?

There was a PR recently that allowed admins to lock papers from the admin fax menu

**Comment by slarticodefast** [comment]
> Many people see papers stamped by Centcom as messages by ADMINS. If the stamped paper said "kill X", the chance is high that someone does it even if it goes 100% against rules.

If that is the way it is intended we should probably remove or lock the Centcom stamped papers (like the PA instructions) some maps have lying around at round start. I used a cybersun pen on those before to gain free access from HOP. I consider this a fun feature tough.

**Comment by chavonadelal** [comment]
> > Many people see papers stamped by Centcom as messages by ADMINS. If the stamped paper said "kill X", the chance is high that someone does it even if it goes 100% against rules.
> 
> If that is the way it is intended we should probably remove or lock the Centcom stamped papers (like the PA instructions) some maps have lying around at round start. I used a cybersun pen on those before to gain free access from HOP. I consider this a fun feature tough.

@slarticodefast I think that nothing prevents you from asking the Central Command for confirmation of the document that confuses you.

**Comment by SlamBamActionman** [comment]
> > Nah, if they are editable it would be much more fun. Tricking the captain with a document sent from centcom about law changes could be some fun rp.
> 
> Many people see papers stamped by Centcom as messages by ADMINS. If the stamped paper said "kill X", the chance is high that someone does it even if it goes 100% against rules.

If admins rely on CentCom-stamped faxes to be seen as admin rules, that is *their problem for using faxes for OOC rules information*. At best CentCom-stamps should be seen as being sent by your in-universe corporate overlords.

As someone who makes up a new stupid law every single Captain shift I do, I think this PR is good on paper but is likely to have pitfalls. It is very easy to write a law that is unfun, and the ones written in Space Law have gone through several revisions to ensure that Sec & Command can enforce them without impacting gameplay.

You don't want to have a rule that ruins actual gameplay for people; by all means, ban cigarettes or alcohol or force everyone to wear hardhats indoors. But you can't ban Science from doing artifacts (without good reason), that just screws over players who want to do that part of the game and is no fun. So you have to be very careful what laws are being changed.

**Comment by chavonadelal** [comment]
Update:

- Localization optimized

- Now, after time has passed for adaptation, a confirmation message is sent that the law has indeed come into force

File: Content.Server/StationEvents/Events/SpaceLawChange.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
+using System.Linq;
```
Comment (Tayrtahn) [inline Content.Server/StationEvents/Events/SpaceLawChange.cs:1]: State should not be stored in systems; it needs to be stored in components.

You might also want to look into LocalizedDataset, which would provide an easier way to store the set of laws and to select one.

**Review by Tayrtahn — CHANGES_REQUESTED** [review]

**Comment by SlamBamActionman** [comment]
It's also worth considering that:
- It seems the laws don't have any sentence tied to them, so if you catch someone doing something that has been outlawed there is no guidance on what the punishment should be.
- The message claims the station should hurry up to avoid potential negative consequences. Since there are none, I would reword this into something that pushes Sec to enforce it. E.g. "Security & Command are to ensure the station is fully compliant with the law by the time the adjustment period is over."

**Comment by Catdere** [comment]
> It's also worth considering that:
> 
> * It seems the laws don't have any sentence tied to them, so if you catch someone doing something that has been outlawed there is no guidance on what the punishment should be.
> * The message claims the station should hurry up to avoid potential negative consequences. Since there are none, I would reword this into something that pushes Sec to enforce it. E.g. "Security & Command are to ensure the station is fully compliant with the law by the time the adjustment period is over."

Hi am epic space lawyer person here it does do this. 1-XX is a Minor crime if thats what the "law code" Within the current screenshots.

File: Content.Server/StationEvents/Events/SpaceLawChange.cs (hunk 1 above)
Comment (chavonadelal) [reply Content.Server/StationEvents/Events/SpaceLawChange.cs]: > State should not be stored in systems; it needs to be stored in components.
> 
> You might also want to look into LocalizedDataset, which would provide an easier way to store the set of laws and to select one.

@Tayrtahn I did as you asked, but I ran into a problem. component.SequenceLaws does not save its state between runs. I do not know what to do. I need to store the list between event triggers so that the options are not repeated, but the list is not saved. Could you please tell me how to store the list between runs?

Comment (chavonadelal) [reply Content.Server/StationEvents/Events/SpaceLawChange.cs]: > State should not be stored in systems; it needs to be stored in components.
> 
> You might also want to look into LocalizedDataset, which would provide an easier way to store the set of laws and to select one.

I move the selected law to the end of the list and select a random law only from the first half of the list so the laws will not be repeated for as long as possible but the list does not save state between runs...

Comment (chavonadelal) [reply Content.Server/StationEvents/Events/SpaceLawChange.cs:2]: > State should not be stored in systems; it needs to be stored in components.
> 
> You might also want to look into LocalizedDataset, which would provide an easier way to store the set of laws and to select one.

@Tayrtahn I have made all the necessary changes. Now System.Linq is used only to be able to use the Except method in lists, the necessary data is stored in the component and the messages are stored in a localized dataset. Everything works.

**Comment by chavonadelal** [comment]
I hope I added the final winter events to the event code. Everything should work. If anyone wants, I will be glad to have suggestions for adding message options.

**Comment by Tayrtahn** [comment]
I'm away for the weekend, but I'll give this another review when I get back.

**Comment by chavonadelal** [comment]
> I'm away for the weekend, but I'll give this another review when I get back.

@Tayrtahn ok thank you I will be looking forward to your review

**Comment by retequizzle** [comment]
going to be transparent here and state that even if it seems like a good idea on paper, it's going to be terrible in implementation on average

borrowing an example from a private discussion about this:

>latejoin crewmember shows up after space law bs, let's assume space law is cigs are now contraband
>get arrested for smoking
>ahelp because they (rightfully) think sec is being shitsec and arresting for no reason
>> and let's be honest - sec could literally show the person in question the fax stating the "space law change" and they would still refuse to believe it
>either current round admin has to investigate - or the more likely scenario, an admin has to get on and sanity check 
1) that its not just shitsec
2) that the law change actually occured
>> and let's be honest - people are already under the impression that their ahelps regularly go unheard because admin presence is stretched incredibly thin

in practice, making random things an arrestable offense doesn't serve to make anything more fun for any party involved. random player feels like they're being metagrudged by sec, sec has to put up with even more harassment from people than they already do, and admins have to spend more time looking into non-issues that absolutely should not be an issue in the first place. 

as such, even if @Tayrtahn finds no further issues on review after their return from their weekend trip, i'm not in favor of this being merged in any capacity upstream. the only reason i'm not closing the PR outright is because i'm typically not a reviewer for these sorts of things and would like to get other PM opinions on the matter, since this extends beyond just a fun player feature by a significant margin.

tl;dr - please leave random amendable laws to borgs and non-organics, no one who logs into the game is sane as is and this would just exacerbate the issue

**Comment by chavonadelal** [comment]
> going to be transparent here and state that even if it seems like a good idea on paper, it's going to be terrible in implementation on average
> 
> borrowing an example from a private discussion about this:
> 
> * latejoin crewmember shows up after space law bs, let's assume space law is cigs are now contraband
> * get arrested for smoking
> * ahelp because they (rightfully) think sec is being shitsec and arresting for no reason. and let's be honest - sec could literally show the person in question the fax stating the "space law change" and they would still refuse to believe it
> * either current round admin has to investigate - or the more likely scenario, an admin has to get on and sanity check:
> 
> 1. that its not just shitsec
> 2. that the law change actually occured
> 
> * which poses additional issues because people are already under the impression that their ahelps regularly go unheard, because admin presence is stretched incredibly thin
> 
> in practice, making random things an arrestable offense doesn't serve to make anything more fun for any party involved. random player feels like they're being metagrudged by sec, sec has to put up with even more harassment from people than they already do, and admins have to spend more time looking into non-issues that absolutely should not be an issue in the first place.
> 
> as such, even if @Tayrtahn finds no further issues on review after their return from their weekend trip, i'm not in favor of this being merged in any capacity upstream. the only reason i'm not closing the PR outright is because i'm typically not a reviewer for these sorts of things and would like to get other PM opinions on the matter, since this extends beyond just a fun player feature by a significant margin.
> 
> tl;dr - please leave random amendable laws to borgs and non-organics, no one who logs into the game is sane as is and this would just exacerbate the issue

I think that the problem you mentioned relates exclusively to the state of administration on the server. If necessary, I can try to additionally indicate on the fax the time it was sent so that players can check the current time and make sure that the law has come into force. But I don't know how to do it yet. A person will simply look at his PDA and find out whether the time has come for the law to come into force. Well, if he still doesn’t believe after that, then this is clearly not a problem with the event, but a problem with the player himself who doesn’t believe.
I also focus on other servers where the situations you mentioned are much less common (I’ll take your word for it that such situations do exist). Provided that this event is very rare and can occur a maximum of 2 times per round (can be reduced to once), nothing particularly bad can happen and rather brings new roleplay experience. If this pull request is closed, I will be very upset because I put a lot of effort as a person who wrote some code for the first time in my life a week ago. I really think it would add variety to the round.

**Comment by Plykiya** [comment]
wouldn't it be more player-inclusive if it was reported via the news app on the PDA rather than just being faxed to some certain people who then have to diseminate the information to the rest of the crew? that way people can still have access to it if they latejoin

**Comment by chavonadelal** [comment]
> wouldn't it be more player-inclusive if it was reported via the news app on the PDA rather than just being faxed to some certain people who then have to disseminate the information to the rest of the crew? that way people can still have access to it if they latejoin

@Plykiya I could create news about changing the cosmic law and thought about it, but I'm afraid that it would be easy for players to delete it and compromise it

**Comment by lzk228** [comment]
the law on wizden is not very good implemented to change it midround

**Comment by Plykiya** [comment]
> > wouldn't it be more player-inclusive if it was reported via the news app on the PDA rather than just being faxed to some certain people who then have to disseminate the information to the rest of the crew? that way people can still have access to it if they latejoin
> 
> @Plykiya I could create news about changing the cosmic law and thought about it, but I'm afraid that it would be easy for players to delete it and compromise it

people can delete news articles?

**Comment by chavonadelal** [comment]
> > > wouldn't it be more player-inclusive if it was reported via the news app on the PDA rather than just being faxed to some certain people who then have to disseminate the information to the rest of the crew? that way people can still have access to it if they latejoin
> > 
> > 
> > @Plykiya I could create news about changing the cosmic law and thought about it, but I'm afraid that it would be easy for players to delete it and compromise it
> 
> people can delete news articles?

As I remember, yes. Service level access is sufficient

**Comment by Plykiya** [comment]
Sounds like news articles just need a bool to ensure some can't be deleted then, same thing as how some papers can be locked to prevent any tampering

**Comment by chavonadelal** [comment]
> the law on wizden is not very good implemented yet to change it midround



> Sounds like news articles just need a bool to ensure some can't be deleted then, same thing as how some papers can be locked to prevent any tampering

I thought about this and I would try to create a new type of news from an authorized source that cannot be deleted and it will be clear that it is definitely from the Central Command, but I am afraid my skills are not enough. I spent a week making these 158 lines for the event.

**Comment by Tayrtahn** [comment]
Thank you for your contribution! However, after discussion among maintainers, we have decided to close this PR on the grounds that the feature is not desired. Concerns were raised about discoverability of law changes, difficulty for admins, and potential to create undesirable player behavior. 

For these reasons, this PR will be closed. You are however encouraged to continue contributing if you are still interested, and you are free to offer this code to forks as well.

On a personal note, I would like to add that I was impressed by how quickly you were able to respond to code review with improvements - I get the sense that you most have learned a fair bit about our codebase and architecture in the process, so I hope to see your work here again in the future.

**Comment by chavonadelal** [comment]
> Thank you for your contribution! However, after discussion among maintainers, we have decided to close this PR on the grounds that the feature is not desired. Concerns were raised about discoverability of law changes, difficulty for admins, and potential to create undesirable player behavior.
> 
> For these reasons, this PR will be closed. You are however encouraged to continue contributing if you are still interested, and you are free to offer this code to forks as well.
> 
> On a personal note, I would like to add that I was impressed by how quickly you were able to respond to code review with improvements - I get the sense that you must have learned a fair bit about our codebase and architecture in the process, so I hope to see your work here again in the future.

It's a pity I really tried  :( The people in the discord helped me figure out the architecture, for which I thank them! Well, then I'll do localization, probably this will be in demand. Thank you for your words.
